File: packages/redis-streams/src/consumer.ts
/**
 * @fileoverview Redis Streams consumer implementation
 * @module @anyq/redis-streams/consumer
 */

import Redis from 'ioredis';
import {
  BaseConsumer,
  createMessage,
  type MessageHandler,
  type BatchMessageHandler,
  type SubscribeOptions,
  type HealthStatus,
  type IMessage,
  type MessageHeaders,
  ConnectionError,
} from '@anyq/core';
import type { RedisStreamsConfig, ConsumerGroupConfig } from './config.js';

/**
 * Parsed stream entry
 */
interface StreamEntry {
  id: string;
  fields: Record<string, string>;
}

/**
 * Default subscribe options
 */
const DEFAULT_SUBSCRIBE_OPTIONS: Required<SubscribeOptions> = {
  fromBeginning: false,
  fromTimestamp: undefined as unknown as Date,
  concurrency: 1,
  autoAck: true,
  batchSize: 10,
  batchTimeout: 1000,
};

/**
 * Redis Streams consumer implementation
 *
 * Consumes messages from Redis Streams using consumer groups.
 *
 * @example
 * ```typescript
 * const consumer = new RedisStreamsConsumer({
 *   driver: 'redis-streams',
 *   streamName: 'orders',
 *   consumerGroup: {
 *     groupName: 'order-processors',
 *     consumerName: 'consumer-1',
 *   },
 *   redis: { host: 'localhost', port: 6379 },
 * });
 *
 * await consumer.connect();
 * await consumer.subscribe(async (message) => {
 *   console.log('Received:', message.body);
 *   await message.ack();
 * });
 * ```
 */
export class RedisStreamsConsumer<T = unknown> extends BaseConsumer<T> {
  private client: Redis | null = null;
  private readonly streamName: string;
  private readonly groupConfig: ConsumerGroupConfig;
  private readonly blockTimeout: number;
  private readonly claimTimeout: number;
  private readonly minIdleTime: number;
  private running = false;
  private pollInterval: ReturnType<typeof setInterval> | null = null;

  constructor(config: RedisStreamsConfig) {
    super(config);
    this.streamName = config.streamName;
    this.groupConfig = config.consumerGroup ?? {
      groupName: 'default-group',
      consumerName: `consumer-${Date.now()}`,
    };
    this.blockTimeout = config.blockTimeout ?? 5000;
    this.claimTimeout = config.claimTimeout ?? 30000;
    this.minIdleTime = config.minIdleTime ?? 10000;
  }

  /**
   * Connect to Redis and create consumer group
   */
  async connect(): Promise<void> {
    if (this._connected) {
      return;
    }

    try {
      const redisConfig = (this.config as RedisStreamsConfig).redis ?? {};

      if (redisConfig.url) {
        this.client = new Redis(redisConfig.url, {
          keyPrefix: redisConfig.keyPrefix,
          connectionName: redisConfig.connectionName,
          lazyConnect: true,
        });
      } else {
        this.client = new Redis({
          host: redisConfig.host ?? 'localhost',
          port: redisConfig.port ?? 6379,
          password: redisConfig.password,
          db: redisConfig.db ?? 0,
          username: redisConfig.username,
          tls: redisConfig.tls ? {} : undefined,
          keyPrefix: redisConfig.keyPrefix,
          connectionName: redisConfig.connectionName,
          lazyConnect: true,
        });
      }

      await this.client.connect();

      // Create consumer group if needed
      if (this.groupConfig.autoCreate !== false) {
        await this.ensureConsumerGroup();
      }

      this._connected = true;
      this.logger.info('Redis Streams consumer connected', {
        stream: this.streamName,
        group: this.groupConfig.groupName,
        consumer: this.groupConfig.consumerName,
      });
    } catch (error) {
      throw new ConnectionError(
        'Failed to connect to Redis',
        error instanceof Error ? error : undefined
      );
    }
  }

  /**
   * Disconnect from Redis
   */
  async disconnect(): Promise<void> {
    if (!this._connected || !this.client) {
      return;
    }

    this.running = false;
    if (this.pollInterval) {
      clearInterval(this.pollInterval);
      this.pollInterval = null;
    }

    try {
      await this.client.quit();
    } catch {
      // Ignore quit errors
    }

    this.client = null;
    this._connected = false;
    this.logger.info('Redis Streams consumer disconnected');
  }

  /**
   * Subscribe to messages
   */
  async subscribe(
    handler: MessageHandler<T>,
    options?: SubscribeOptions
  ): Promise<void> {
    if (!this.client) {
      throw new ConnectionError('Consumer not connected');
    }

    const opts = { ...DEFAULT_SUBSCRIBE_OPTIONS, ...options };
    this.running = true;

    const poll = async () => {
      if (!this.running || this._paused || !this.client) {
        return;
      }

      try {
        // First, try to claim pending messages
        if (this.claimTimeout > 0) {
          await this.processPendingMessages(handler, opts);
        }

        // Then read new messages
        const entries = await this.readMessages(1);

        for (const entry of entries) {
          if (!this.running || this._paused) break;

          const message = this.createWrappedMessage(entry);

          try {
            this.emit('message', message);
            await handler(message);

            if (opts.autoAck) {
              await message.ack();
            }
          } catch (error) {
            this.logger.error('Error processing message', {
              messageId: entry.id,
              error: error instanceof Error ? error.message : String(error),
            });

            this.emit(
              'error',
              error instanceof Error ? error : new Error(String(error))
            );

            // Let the message become pending for reclaim
          }
        }
      } catch (error) {
        this.logger.error('Error in poll loop', {
          error: error instanceof Error ? error.message : String(error),
        });
      }
    };

    // Start polling
    this.pollInterval = setInterval(poll, 100);
    await poll(); // Initial poll

    this.logger.info('Subscribed to stream', {
      stream: this.streamName,
      group: this.groupConfig.groupName,
    });
  }

  /**
   * Subscribe to message batches
   */
  async subscribeBatch(
    handler: BatchMessageHandler<T>,
    options?: SubscribeOptions
  ): Promise<void> {
    if (!this.client) {
      throw new ConnectionError('Consumer not connected');
    }

    const opts = { ...DEFAULT_SUBSCRIBE_OPTIONS, ...options };
    this.running = true;

    const poll = async () => {
      if (!this.running || this._paused || !this.client) {
        return;
      }

      try {
        const entries = await this.readMessages(opts.batchSize);

        if (entries.length === 0) {
          return;
        }

        const messages = entries.map((entry) => this.createWrappedMessage(entry));

        try {
          await handler(messages);

          if (opts.autoAck) {
            for (const message of messages) {
              await message.ack();
            }
          }
        } catch (error) {
          this.logger.error('Error processing batch', {
            count: messages.length,
            error: error instanceof Error ? error.message : String(error),
          });

          this.emit(
            'error',
            error instanceof Error ? error : new Error(String(error))
          );
        }
      } catch (error) {
        this.logger.error('Error in batch poll loop', {
          error: error instanceof Error ? error.message : String(error),
        });
      }
    };

    this.pollInterval = setInterval(poll, opts.batchTimeout);
    this.logger.info('Subscribed to stream (batch)', {
      stream: this.streamName,
      group: this.groupConfig.groupName,
    });
  }

  /**
   * Pause consumption
   */
  async pause(): Promise<void> {
    await super.pause();
    this.running = false;
  }

  /**
   * Resume consumption
   */
  async resume(): Promise<void> {
    await super.resume();
    this.running = true;
  }

  /**
   * Get health status
   */
  async healthCheck(): Promise<HealthStatus> {
    const start = Date.now();

    try {
      if (!this.client || !this._connected) {
        return {
          healthy: false,
          connected: false,
          error: 'Not connected',
        };
      }

      await this.client.ping();

      // Get consumer group info
      const groups = await this.client.xinfo('GROUPS', this.streamName).catch(() => []);
      const groupInfo = (groups as unknown[][]).find(
        (g) => g[g.indexOf('name') + 1] === this.groupConfig.groupName
      );

      return {
        healthy: true,
        connected: true,
        latencyMs: Date.now() - start,
        details: {
          stream: this.streamName,
          group: this.groupConfig.groupName,
          consumer: this.groupConfig.consumerName,
          pending: groupInfo ? groupInfo[groupInfo.indexOf('pending') + 1] : 0,
          paused: this._paused,
        },
      };
    } catch (error) {
      return {
        healthy: false,
        connected: false,
        latencyMs: Date.now() - start,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  /**
   * Get the underlying Redis client (for testing)
   */
  getClient(): Redis | null {
    return this.client;
  }

  /**
   * Ensure the consumer group exists
   */
  private async ensureConsumerGroup(): Promise<void> {
    if (!this.client) return;

    try {
      const startId = this.groupConfig.startId ?? '0';
      await this.client.xgroup(
        'CREATE',
        this.streamName,
        this.groupConfig.groupName,
        startId,
        'MKSTREAM'
      );
      this.logger.debug('Consumer group created', {
        group: this.groupConfig.groupName,
        stream: this.streamName,
      });
    } catch (error) {
      // Group already exists
      if (!(error instanceof Error) || !error.message.includes('BUSYGROUP')) {
        throw error;
      }
    }
  }

  /**
   * Read messages from the stream
   */
  private async readMessages(count: number): Promise<StreamEntry[]> {
    if (!this.client) return [];

    try {
      const result = await this.client.xreadgroup(
        'GROUP',
        this.groupConfig.groupName,
        this.groupConfig.consumerName,
        'COUNT',
        String(count),
        'BLOCK',
        String(this.blockTimeout),
        'STREAMS',
        this.streamName,
        '>'
      );

      if (!result) return [];

      // Parse the result
      const entries: StreamEntry[] = [];
      for (const [, messages] of result as [string, [string, string[]][]][]) {
        for (const [id, fields] of messages) {
          const fieldMap: Record<string, string> = {};
          for (let i = 0; i < fields.length; i += 2) {
            fieldMap[fields[i]] = fields[i + 1];
          }
          entries.push({ id, fields: fieldMap });
        }
      }

      return entries;
    } catch (error) {
      this.logger.error('Error reading messages', {
        error: error instanceof Error ? error.message : String(error),
      });
      return [];
    }
  }

  /**
   * Process pending messages (auto-claim)
   */
  private async processPendingMessages(
    handler: MessageHandler<T>,
    opts: Required<SubscribeOptions>
  ): Promise<void> {
    if (!this.client) return;

    try {
      // Use XAUTOCLAIM to claim and get pending messages
      const result = await this.client.xautoclaim(
        this.streamName,
        this.groupConfig.groupName,
        this.groupConfig.consumerName,
        String(this.minIdleTime),
        '0-0',
        'COUNT',
        '5'
      );

      if (!result || !Array.isArray(result) || result.length < 2) return;

      const messages = result[1] as [string, string[]][];
      for (const [id, fields] of messages) {
        if (!this.running || this._paused) break;

        const fieldMap: Record<string, string> = {};
        for (let i = 0; i < fields.length; i += 2) {
          fieldMap[fields[i]] = fields[i + 1];
        }

        const message = this.createWrappedMessage({ id, fields: fieldMap });

        try {
          await handler(message);
          if (opts.autoAck) {
            await message.ack();
          }
        } catch (error) {
          this.logger.error('Error processing pending message', {
            messageId: id,
            error: error instanceof Error ? error.message : String(error),
          });
        }
      }
    } catch (error) {
      // XAUTOCLAIM might fail on older Redis versions
      this.logger.debug('XAUTOCLAIM failed, skipping pending processing', {
        error: error instanceof Error ? error.message : String(error),
      });
    }
  }

  /**
   * Create a wrapped message from stream entry
   */
  private createWrappedMessage(entry: StreamEntry): IMessage<T> {
    const body = this.serializer.deserialize(entry.fields.body || '{}');
    const headers: MessageHeaders = entry.fields.headers
      ? JSON.parse(entry.fields.headers)
      : {};

    return createMessage({
      id: entry.id,
      body,
      key: entry.fields.key,
      headers,
      timestamp: new Date(parseInt(entry.id.split('-')[0], 10)),
      deliveryAttempt: 1,
      metadata: {
        provider: 'redis-streams',
        redisStreams: {
          stream: this.streamName,
          entryId: entry.id,
          consumerGroup: this.groupConfig.groupName,
          consumer: this.groupConfig.consumerName,
        },
      },
      raw: entry,
      onAck: async () => {
        if (this.client) {
          await this.client.xack(
            this.streamName,
            this.groupConfig.groupName,
            entry.id
          );
          this.logger.debug('Message acknowledged', { messageId: entry.id });
        }
      },
      onNack: async (requeue = true) => {
        // For Redis Streams, nack doesn't have a direct equivalent
        // If requeue=false, we remove from pending list
        if (!requeue && this.client) {
          await this.client.xack(
            this.streamName,
            this.groupConfig.groupName,
            entry.id
          );
        }
        this.logger.debug('Message nacked', { messageId: entry.id, requeue });
      },
    });
  }
}

File: go/redis/consumer.go
package redis

import (
	"context"
	"encoding/json"
	"strconv"
	"strings"
	"time"

	goredis "github.com/redis/go-redis/v9"
	"github.com/sns45/anyq/go/core"
)

// Consumer reads messages from a Redis Stream via a consumer group (XREADGROUP)
// and acknowledges with XACK. It mirrors the TypeScript RedisStreamsConsumer.
//
// Capability note: Redis Streams has no native delayed-redelivery primitive, so
// SupportsNativeDelay is left false (the base default). Park decisions therefore
// downgrade to in-process retry inside core.ApplyStrategy.
type Consumer struct {
	*core.BaseConsumer

	cfg          Config
	streamName   string
	groupName    string
	consumerName string
	blockTimeout time.Duration
	claimTimeout time.Duration
	minIdleTime  time.Duration

	client *goredis.Client
}

// NewConsumer builds a Redis Streams consumer. Like the memory reference, it
// embeds *core.BaseConsumer, calls InitBase then Bind(self) so ApplyStrategy
// dispatches to this adapter's hook overrides.
func NewConsumer(cfg Config) *Consumer {
	cfg = cfg.withDefaults()
	bc := &core.BaseConsumer{}
	bc.InitBase(cfg.BaseQueueConfig)

	consumerName := cfg.ConsumerGroup.ConsumerName
	if consumerName == "" {
		consumerName = "consumer-" + strconv.FormatInt(time.Now().UnixMilli(), 10)
	}

	c := &Consumer{
		BaseConsumer: bc,
		cfg:          cfg,
		streamName:   cfg.Redis.KeyPrefix + cfg.StreamName,
		groupName:    cfg.ConsumerGroup.GroupName,
		consumerName: consumerName,
		blockTimeout: time.Duration(cfg.BlockTimeoutMs) * time.Millisecond,
		claimTimeout: time.Duration(cfg.ClaimTimeoutMs) * time.Millisecond,
		minIdleTime:  time.Duration(cfg.MinIdleTimeMs) * time.Millisecond,
	}
	bc.Bind(c)
	return c
}

// Connect dials Redis and creates the consumer group (XGROUP CREATE MKSTREAM).
func (c *Consumer) Connect(ctx context.Context) error {
	if c.IsConnected() {
		return nil
	}
	if err := c.VerifyParkPolicy(); err != nil {
		return err
	}
	client, err := newClient(c.cfg.Redis)
	if err != nil {
		return core.NewConnectionError("failed to connect to Redis", err)
	}
	if err := client.Ping(ctx).Err(); err != nil {
		_ = client.Close()
		return core.NewConnectionError("failed to connect to Redis", err)
	}
	c.client = client

	if c.cfg.ConsumerGroup.autoCreateEnabled() {
		if err := c.ensureConsumerGroup(ctx); err != nil {
			_ = client.Close()
			c.client = nil
			return core.NewConnectionError("failed to connect to Redis", err)
		}
	}

	c.SetConnected(true)
	c.Logger.Info("Redis Streams consumer connected", map[string]any{
		"stream": c.streamName, "group": c.groupName, "consumer": c.consumerName,
	})
	return nil
}

// ensureConsumerGroup creates the group, ignoring BUSYGROUP (already exists).
func (c *Consumer) ensureConsumerGroup(ctx context.Context) error {
	startID := c.cfg.ConsumerGroup.StartID
	if startID == "" {
		startID = defaultStartID
	}
	err := c.client.XGroupCreateMkStream(ctx, c.streamName, c.groupName, startID).Err()
	if err != nil && !strings.Contains(err.Error(), "BUSYGROUP") {
		return err
	}
	c.Logger.Debug("consumer group ready", map[string]any{"group": c.groupName, "stream": c.streamName})
	return nil
}

// Disconnect closes the Redis connection.
func (c *Consumer) Disconnect(ctx context.Context) error {
	if !c.IsConnected() || c.client == nil {
		return nil
	}
	_ = c.client.Close()
	c.client = nil
	c.SetConnected(false)
	c.Logger.Info("Redis Streams consumer disconnected", nil)
	return nil
}

// Subscribe consumes messages until ctx is cancelled. On handler success it
// XACKs; on handler error it runs core.ApplyStrategy with a reinvoke closure,
// and on the !handled fallback preserves the TS legacy behaviour (log + emit
// error; leave the entry pending for later reclaim).
func (c *Consumer) Subscribe(ctx context.Context, handler core.Handler, opts *core.SubscribeOptions) error {
	if c.client == nil {
		return core.NewConnectionError("consumer not connected", nil)
	}
	autoAck := opts.AutoAckEnabled()
	c.Logger.Info("subscribed to stream", map[string]any{"stream": c.streamName, "group": c.groupName})

	for {
		if err := ctx.Err(); err != nil {
			return err
		}
		if c.IsPaused() {
			if err := sleepCtx(ctx, 100*time.Millisecond); err != nil {
				return err
			}
			continue
		}

		// Reclaim idle pending messages first (parity with TS poll loop).
		if c.claimTimeout > 0 {
			if err := c.processPending(ctx, handler, autoAck); err != nil {
				return err
			}
		}

		entries, err := c.readMessages(ctx, 1)
		if err != nil {
			return err
		}
		for _, entry := range entries {
			if err := ctx.Err(); err != nil {
				return err
			}
			if c.IsPaused() {
				break
			}
			if herr := c.handleOne(ctx, handler, entry, autoAck); herr != nil {
				return herr
			}
		}
	}
}

// handleOne processes a single entry, returning a non-nil error only for fatal
// outcomes (Fail decision / context cancellation) that should stop the loop.
func (c *Consumer) handleOne(ctx context.Context, handler core.Handler, entry goredis.XMessage, autoAck bool) error {
	msg := c.wrap(entry)
	herr := handler(ctx, msg)
	if herr == nil {
		if autoAck {
			return msg.Ack(ctx)
		}
		return nil
	}

	reinvoke := func() error { return handler(ctx, msg) }
	handled, applyErr := c.ApplyStrategy(ctx, msg, herr, reinvoke)
	if applyErr != nil {
		return applyErr
	}
	if handled {
		return nil
	}

	// Legacy behaviour (TS !result.handled branch): log, emit error, and leave
	// the entry pending so it can be reclaimed later. No ack/nack here.
	c.Logger.Error("error processing message", map[string]any{
		"messageId": entry.ID, "error": herr.Error(),
	})
	if c.Config.OnError != nil {
		c.Config.OnError(herr)
	}
	return nil
}

// processPending reclaims and re-processes idle pending entries via XAUTOCLAIM.
func (c *Consumer) processPending(ctx context.Context, handler core.Handler, autoAck bool) error {
	msgs, _, err := c.client.XAutoClaim(ctx, &goredis.XAutoClaimArgs{
		Stream:   c.streamName,
		Group:    c.groupName,
		Consumer: c.consumerName,
		MinIdle:  c.minIdleTime,
		Start:    "0-0",
		Count:    5,
	}).Result()
	if err != nil {
		// XAUTOCLAIM may be unavailable on older Redis; mirror TS debug-and-skip.
		c.Logger.Debug("XAUTOCLAIM failed; skipping pending processing", map[string]any{"error": err.Error()})
		return nil
	}
	for _, entry := range msgs {
		if err := ctx.Err(); err != nil {
			return err
		}
		if c.IsPaused() {
			break
		}
		if herr := c.handleOne(ctx, handler, entry, autoAck); herr != nil {
			return herr
		}
	}
	return nil
}

// SubscribeBatch consumes batches; on batch error applies strategy per message
// (Redis Streams supports per-entry ack), matching the TS adapter.
func (c *Consumer) SubscribeBatch(ctx context.Context, handler core.BatchHandler, opts *core.SubscribeOptions) error {
	if c.client == nil {
		return core.NewConnectionError("consumer not connected", nil)
	}
	batchSize := c.cfg.BatchSize
	if opts != nil && opts.BatchSize > 0 {
		batchSize = opts.BatchSize
	}
	autoAck := opts.AutoAckEnabled()
	c.Logger.Info("subscribed to stream (batch)", map[string]any{"stream": c.streamName, "group": c.groupName})

	for {
		if err := ctx.Err(); err != nil {
			return err
		}
		if c.IsPaused() {
			if err := sleepCtx(ctx, 100*time.Millisecond); err != nil {
				return err
			}
			continue
		}

		entries, err := c.readMessages(ctx, int64(batchSize))
		if err != nil {
			return err
		}
		if len(entries) == 0 {
			continue
		}
		msgs := make([]core.Message, len(entries))
		for i, e := range entries {
			msgs[i] = c.wrap(e)
		}

		herr := handler(ctx, msgs)
		if herr == nil {
			if autoAck {
				for _, m := range msgs {
					_ = m.Ack(ctx)
				}
			}
			continue
		}

		anyUnhandled := false
		for _, m := range msgs {
			mm := m
			reinvoke := func() error { return handler(ctx, []core.Message{mm}) }
			handled, applyErr := c.ApplyStrategy(ctx, mm, herr, reinvoke)
			if applyErr != nil {
				return applyErr
			}
			if !handled {
				anyUnhandled = true
			}
		}
		if anyUnhandled {
			c.Logger.Error("error processing batch", map[string]any{"count": len(msgs), "error": herr.Error()})
			if c.Config.OnError != nil {
				c.Config.OnError(herr)
			}
		}
	}
}

// readMessages reads up to count new entries via XREADGROUP with BLOCK.
func (c *Consumer) readMessages(ctx context.Context, count int64) ([]goredis.XMessage, error) {
	streams, err := c.client.XReadGroup(ctx, &goredis.XReadGroupArgs{
		Group:    c.groupName,
		Consumer: c.consumerName,
		Streams:  []string{c.streamName, ">"},
		Count:    count,
		Block:    c.blockTimeout,
	}).Result()
	if err != nil {
		if err == goredis.Nil {
			return nil, nil // BLOCK timeout, no messages
		}
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}
		// Transient read error: mirror TS log-and-return-empty so the loop retries.
		c.Logger.Error("error reading messages", map[string]any{"error": err.Error()})
		return nil, nil
	}
	var out []goredis.XMessage
	for _, s := range streams {
		out = append(out, s.Messages...)
	}
	return out, nil
}

// Pause and Resume rely on the embedded BaseConsumer defaults; the Subscribe
// loops check IsPaused().

// HealthCheck pings Redis and reports consumer-group pending count.
func (c *Consumer) HealthCheck(ctx context.Context) (core.HealthStatus, error) {
	start := time.Now()
	if c.client == nil || !c.IsConnected() {
		return core.HealthStatus{Healthy: false, Connected: false, Error: "Not connected"}, nil
	}
	if err := c.client.Ping(ctx).Err(); err != nil {
		return core.HealthStatus{
			Healthy:   false,
			Connected: false,
			LatencyMs: time.Since(start).Milliseconds(),
			Error:     err.Error(),
		}, nil
	}
	var pending int64
	if groups, err := c.client.XInfoGroups(ctx, c.streamName).Result(); err == nil {
		for _, g := range groups {
			if g.Name == c.groupName {
				pending = g.Pending
				break
			}
		}
	}
	return core.HealthStatus{
		Healthy:   true,
		Connected: true,
		LatencyMs: time.Since(start).Milliseconds(),
		Details: map[string]any{
			"stream":   c.streamName,
			"group":    c.groupName,
			"consumer": c.consumerName,
			"pending":  pending,
			"paused":   c.IsPaused(),
		},
	}, nil
}

// Client returns the underlying Redis client (for testing).
func (c *Consumer) Client() *goredis.Client { return c.client }

// wrap builds a core.Message from a Redis stream entry. Ack performs XACK; Nack
// performs XACK only when requeue is false (remove from pending), matching TS.
func (c *Consumer) wrap(entry goredis.XMessage) core.Message {
	body := fieldString(entry.Values, "body")
	key := fieldString(entry.Values, "key")

	headers := core.MessageHeaders{}
	if raw := fieldString(entry.Values, "headers"); raw != "" {
		var decoded map[string]string
		if err := json.Unmarshal([]byte(raw), &decoded); err == nil {
			for k, v := range decoded {
				headers[k] = []byte(v)
			}
		}
	}

	return core.NewMessage(core.MessageParams{
		ID:              entry.ID,
		Body:            []byte(body),
		Key:             key,
		Headers:         headers,
		Timestamp:       timestampFromID(entry.ID),
		DeliveryAttempt: 1,
		Metadata: core.ProviderMetadata{
			Provider: core.DriverRedisStreams,
			RedisStreams: &core.RedisStreamsMetadata{
				Stream:        c.streamName,
				EntryID:       entry.ID,
				ConsumerGroup: c.groupName,
				Consumer:      c.consumerName,
			},
		},
		Raw: entry,
		OnAck: func(ctx context.Context) error {
			if c.client == nil {
				return nil
			}
			if err := c.client.XAck(ctx, c.streamName, c.groupName, entry.ID).Err(); err != nil {
				return err
			}
			c.Logger.Debug("message acknowledged", map[string]any{"messageId": entry.ID})
			return nil
		},
		OnNack: func(ctx context.Context, requeue bool) error {
			// No direct nack in Redis Streams. requeue=false => XACK (drop from
			// pending). requeue=true => leave pending for reclaim.
			if !requeue && c.client != nil {
				if err := c.client.XAck(ctx, c.streamName, c.groupName, entry.ID).Err(); err != nil {
					return err
				}
			}
			c.Logger.Debug("message nacked", map[string]any{"messageId": entry.ID, "requeue": requeue})
			return nil
		},
	})
}

// fieldString reads a string field from the XADD value map.
func fieldString(values map[string]interface{}, key string) string {
	if v, ok := values[key]; ok {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}

// timestampFromID derives a timestamp from the millis portion of an entry id.
func timestampFromID(id string) time.Time {
	ms := id
	if i := strings.IndexByte(id, '-'); i >= 0 {
		ms = id[:i]
	}
	n, err := strconv.ParseInt(ms, 10, 64)
	if err != nil {
		return time.Now()
	}
	return time.UnixMilli(n)
}

// sleepCtx sleeps for d or until ctx is cancelled, returning ctx.Err() on cancel.
func sleepCtx(ctx context.Context, d time.Duration) error {
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-t.C:
		return nil
	}
}

var _ core.Consumer = (*Consumer)(nil)
var _ core.ConsumerHooks = (*Consumer)(nil)
